File: services/data_gateway/app/routers/code_evidence.py
"""Code quality + similarity evidence — PH4-D3.

``hr_router`` only (``/hr``, ``HrCtxDep``): only ``hr_manager`` ever reads a
named candidate's source or a similarity comparison — both audited every
time (D3 #24) — or records an integrity finding. There is no super-admin,
interviewer or agent route, and no candidate route at all.
"""

from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel, Field
from sqlalchemy import text

from app import code_evidence as svc
from app.config import settings
from app.database import DbSessionDep
from app.dependencies import HrCtxDep
from app.interviewer_scorecards import RequestMeta
from app.models import Enrolment, Exam, ExamAttempt
from app.rate_limit import rate_limit_company
from app.utils.ownership import get_owned
from app.utils.request_ip import extract_client_ip, extract_user_agent

hr_router = APIRouter(prefix="/hr", tags=["code-evidence"])


def _meta(request: Request) -> RequestMeta:
    return RequestMeta(ip_address=extract_client_ip(request), user_agent=extract_user_agent(request))


async def _fail(exc: svc.CodeEvidenceError) -> HTTPException:
    return HTTPException(status_code=exc.status_code, detail=exc.detail)


async def _owned_attempt(db: DbSessionDep, company_id: uuid.UUID, exam_id: uuid.UUID, aid: uuid.UUID) -> ExamAttempt:
    await get_owned(db, Exam, company_id, exam_id, noun="Exam")
    return await get_owned(db, ExamAttempt, company_id, aid, noun="Attempt", exam_id=exam_id)


async def _enrolment_for_attempt(db: DbSessionDep, company_id: uuid.UUID, attempt_id: uuid.UUID) -> uuid.UUID | None:
    row = await db.scalar(
        text(
            "SELECT asg.enrolment_id FROM exam_attempts a"
            "  JOIN exam_assignments asg ON asg.id = a.assignment_id AND asg.company_id = a.company_id"
            " WHERE a.id = :a AND a.company_id = :c"
        ),
        {"a": attempt_id, "c": company_id},
    )
    return row


# ---------------------------------------------------------------------------
# Bodies
# ---------------------------------------------------------------------------
class FindingIn(BaseModel):
    attempt_id: uuid.UUID
    coding_question_id: uuid.UUID
    outcome: str = Field(pattern="^(no_concern|follow_up|confirmed)$")
    rationale: str = Field(min_length=20, max_length=2000)
    signal_id: uuid.UUID | None = None
    supersedes_id: uuid.UUID | None = None


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------
@hr_router.get("/exams/{exam_id}/attempts/{aid}/code-evidence")
async def get_code_evidence(
    exam_id: uuid.UUID, aid: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep, request: Request,
) -> dict[str, Any]:
    """MEDIUM-3: audited every time — ``test_results`` carries the
    candidate's program stdout/stderr, the same kind of read source and
    compare views are already audited for (D3 #24)."""
    uid, company_id = ctx
    await _owned_attempt(db, company_id, exam_id, aid)
    try:
        return await svc.evidence_for_attempt(
            db, company_id=company_id, exam_id=exam_id, attempt_id=aid,
            actor=uid, meta=_meta(request),
        )
    except svc.CodeEvidenceError as exc:
        raise await _fail(exc) from exc


@hr_router.get("/exams/{exam_id}/attempts/{aid}/code/{question_id}")
async def get_code_source(
    exam_id: uuid.UUID, aid: uuid.UUID, question_id: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep,
    request: Request,
) -> dict[str, Any]:
    """The candidate's own submitted source — audited every time (D3 #24)."""
    uid, company_id = ctx
    await _owned_attempt(db, company_id, exam_id, aid)
    try:
        return await svc.source_for(
            db, company_id=company_id, exam_id=exam_id, attempt_id=aid, coding_question_id=question_id,
            actor=uid, meta=_meta(request),
        )
    except svc.CodeEvidenceError as exc:
        raise await _fail(exc) from exc


@hr_router.post(
    "/exams/{exam_id}/attempts/{aid}/code-analysis",
    dependencies=[rate_limit_company("code_analysis_ondemand", settings.code_analysis_ondemand_per_minute)],
)
async def trigger_code_analysis(
    exam_id: uuid.UUID, aid: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep,
) -> dict[str, int]:
    """On-demand analysis for one attempt — the sweep already covers recent
    submissions; this is for an older one, or a re-run.

    MEDIUM-4: rate-limited per COMPANY, not just guarded by the sandbox's own
    concurrency cap — each call runs a full pairwise comparison pass for
    every coding question the attempt touches, so an unthrottled HR console
    (or a scripted loop against it) could still repeat that cost far more
    often than the nightly sweep ever would.
    """
    _uid, company_id = ctx
    await _owned_attempt(db, company_id, exam_id, aid)
    try:
        result = await svc.analyse_attempt(db, company_id=company_id, attempt_id=aid)
    except svc.CodeEvidenceError as exc:
        raise await _fail(exc) from exc
    return {
        "attempts_scanned": result.attempts_scanned, "reports_written": result.reports_written,
        "fingerprints_written": result.fingerprints_written, "signals_written": result.signals_written,
    }


@hr_router.get("/enrolments/{enrolment_id}/code-evidence-summary")
async def get_enrolment_code_evidence_summary(
    enrolment_id: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep,
) -> dict[str, int]:
    """The one-enrolment read the candidate drawer calls — counts only, no
    source, no content, no names. The decision queue gets the same counts in
    bulk via ``summary_for_enrolments`` wired into ``get_decision_queue``
    (``app/routers/hr_workflows.py``), exactly as ``interviewer_scorecards``'s
    own summary is."""
    _uid, company_id = ctx
    await get_owned(db, Enrolment, company_id, enrolment_id, noun="Application")
    counts = await svc.summary_for_enrolments(db, company_id=company_id, enrolment_ids=[enrolment_id])
    return counts.get(str(enrolment_id)) or dict(svc.EMPTY_SUMMARY)


@hr_router.get("/exams/{exam_id}/similarity")
async def list_similarity_signals(
    exam_id: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep,
) -> list[dict[str, Any]]:
    _uid, company_id = ctx
    await get_owned(db, Exam, company_id, exam_id, noun="Exam")
    return await svc.signals_for_exam(db, company_id=company_id, exam_id=exam_id)


@hr_router.get("/code-similarity/{signal_id}")
async def get_similarity_compare(
    signal_id: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep, request: Request,
) -> dict[str, Any]:
    """Excerpts of each matched region — audited every time (D3 #24)."""
    uid, company_id = ctx
    try:
        return await svc.compare_view(db, company_id=company_id, signal_id=signal_id, actor=uid,
                                      meta=_meta(request))
    except svc.CodeEvidenceError as exc:
        raise await _fail(exc) from exc


@hr_router.post("/code-integrity-findings", status_code=201)
async def record_code_integrity_finding(
    body: FindingIn, ctx: HrCtxDep, db: DbSessionDep, request: Request,
) -> dict[str, str]:
    uid, company_id = ctx
    await get_owned(db, ExamAttempt, company_id, body.attempt_id, noun="Attempt")
    enrolment_id = await _enrolment_for_attempt(db, company_id, body.attempt_id)
    try:
        fid = await svc.record_finding(
            db, company_id=company_id, attempt_id=body.attempt_id,
            coding_question_id=body.coding_question_id, outcome=body.outcome,
            rationale=body.rationale, actor=uid, meta=_meta(request), signal_id=body.signal_id,
            enrolment_id=enrolment_id, supersedes_id=body.supersedes_id,
        )
    except svc.CodeEvidenceError as exc:
        await db.rollback()
        raise await _fail(exc) from exc
    return {"id": str(fid)}


@hr_router.get("/exams/{exam_id}/attempts/{aid}/integrity-findings")
async def list_integrity_findings(
    exam_id: uuid.UUID, aid: uuid.UUID, ctx: HrCtxDep, db: DbSessionDep, request: Request,
    include_redacted: bool = Query(default=False),
) -> list[dict[str, Any]]:
    uid, company_id = ctx
    await _owned_attempt(db, company_id, exam_id, aid)
    evidence = await svc.evidence_for_attempt(
        db, company_id=company_id, exam_id=exam_id, attempt_id=aid, actor=uid, meta=_meta(request),
    )
    findings = evidence["findings"]
    if not include_redacted:
        findings = [f for f in findings if not f["redacted"]]
    return findings

File: web/src/api/codeEvidence.ts
// codeEvidence.ts — PH4-D3 code quality + similarity EVIDENCE. HR only — every
// route lives under services/data_gateway/app/routers/code_evidence.py's
// `hr_router` (`/hr`, HrCtxDep). There is no super-admin, interviewer,
// candidate or agent route for any of this.
//
// A SIGNAL IS NEVER A FINDING. `code_quality_reports`, `code_fingerprints` and
// `code_similarity_signals` are SYSTEM evidence, written only by the sweep or
// an on-demand re-run, and immutable once written. `code_integrity_findings`
// is the only HUMAN evidence — a named `hr_manager`'s judgement call, always
// with a mandatory rationale. Screens built against this module must keep
// those two kinds visually and structurally separate; see
// components/hr/CodeEvidencePanel.tsx and CodeSimilarityCompare.tsx.
//
// Test COVERAGE is not available for any language — `coverage.available` is
// always `false`. Never render a coverage figure; the sandbox's own TEST
// RESULTS (`test_results`, from `graded_snapshot`) are shown instead, labelled
// as results, not coverage.
//
// Viewing a candidate's source, or a similarity comparison's excerpts, is
// audited server-side on every call (D3 #24) — callers must not prefetch
// either before HR has actually asked to see it.
//
// Field names match app/code_evidence.py's `evidence_for_attempt`,
// `compare_view`, `signals_for_exam` and `record_finding` EXACTLY.

import type { CodingResult } from './exams';
import { apiGet, apiPost } from './client';
import { pathId } from './pathId';

// ---------------------------------------------------------------------------
// Quality — reports (SYSTEM evidence)
// ---------------------------------------------------------------------------

export type AnalyserKind = 'python-ast' | 'pygments-tokens';
export type ReportStatus = 'complete' | 'unsupported' | 'failed' | 'skipped';

export interface CodeFinding {
  rule: string;
  severity: 'info' | 'warning';
  line: number;
  message: string;
}

export interface HalsteadMetrics {
  distinct_operators: number;
  distinct_operands: number;
  total_operators: number;
  total_operands: number;
  vocabulary: number;
  length: number;
  volume: number;
  difficulty: number;
  effort: number;
}

/** Python only — real, per-function metrics from the AST. */
export interface PythonFunctionMetric {
  name: string;
  line: number;
  cyclomatic_complexity: number;
  nesting_depth: number;
  length_lines: number;
  parameter_count: number;
}

export interface PythonMetrics {
  kind: 'python-ast';
  functions: PythonFunctionMetric[];
  function_count: number;
  complexity_total: number;
  complexity_avg: number;
  halstead: HalsteadMetrics;
  maintainability_index: number;
  lines_of_code: number;
  duplication_ratio: number;
}

/** The other nine languages (and Python source `ast.parse` could not read) —
 *  a TOKEN-LEVEL APPROXIMATION. Never present this as equal to `PythonMetrics`. */
export interface TokenMetrics {
  kind: 'token-approximate';
  decision_points: number;
  decision_point_density: number;
  max_nesting_depth: number;
  comment_ratio: number;
  long_line_count: number;
  lines_of_code: number;
  halstead: HalsteadMetrics;
  maintainability_index: number;
  duplication_ratio: number;
}

/** `{}` for `unsupported` / `failed` / `skipped` reports. */
export type CodeQualityMetrics = PythonMetrics | TokenMetrics | Record<string, never>;

/** Always `available: false` — coverage needs instrumented execution, which
 *  this analyser (pure static analysis) never does. Never render a figure
 *  here; show `test_results` instead. */
export interface CoverageUnavailable {
  available: false;
  reason: string;
}

export interface CodeQualityReport {
  coding_question_id: string;
  language: string;
  analyser: AnalyserKind;
  analyser_version: string;
  status: ReportStatus;
  metrics: CodeQualityMetrics;
  findings: CodeFinding[];
  coverage: CoverageUnavailable;
  error_class: string | null;
  source_sha256: string;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Integrity — the proctoring summary already captured at submit time
// ---------------------------------------------------------------------------

export interface ProctoringSummary {
  counts?: Record<string, number>;
  violations?: number;
}

export interface CodeEvidenceIntegrity {
  integrity_score: number | null;
  proctoring_summary: ProctoringSummary | null;
  event_counts: Record<string, number>;
}

// ---------------------------------------------------------------------------
// Similarity signals — SYSTEM evidence, automated and unreviewed
// ---------------------------------------------------------------------------

export type SimilarityReferenceKind = 'submission' | 'reference_solution';

/** One signal as it appears on an attempt's own evidence — carries the fixed
 *  caption server-side, verbatim: similar code is not evidence of misconduct
 *  on its own. */
export interface AttemptSimilaritySignal {
  id: string;
  coding_question_id: string;
  attempt_low_id: string;
  /** `null` when `reference_kind` is `reference_solution` — the "other side"
   *  is the question's reference solution, not another attempt. */
  attempt_high_id: string | null;
  reference_kind: SimilarityReferenceKind;
  containment_low: number;
  containment_high: number | null;
  jaccard: number;
  shared_fingerprints: number;
  tokens_low: number;
  tokens_high: number;
  algorithm_version: string;
  created_at: string;
  caption: string;
}

/** One signal in an exam-wide list (`signals_for_exam`) — no `caption`,
 *  `tokens_*` or `algorithm_version`; render the same fixed caption in the UI
 *  rather than treating its absence as "reviewed". */
export interface ExamSimilaritySignal {
  id: string;
  coding_question_id: string;
  attempt_low_id: string;
  attempt_high_id: string | null;
  reference_kind: SimilarityReferenceKind;
  containment_low: number;
  containment_high: number | null;
  jaccard: number;
  shared_fingerprints: number;
  created_at: string;
}

export interface MatchedRegion {
  low_start: number;
  low_end: number;
  high_start: number;
  high_end: number;
}

/** One contiguous run of source lines, carrying the ABSOLUTE line it starts
 *  on. Number and highlight from `start_line + index`, never from a block's
 *  own position in the array — `matched_regions` are absolute source line
 *  numbers too, and the previous version of this screen numbered rows 1..N
 *  by excerpt position, which only ever lined up while excerpts were the
 *  first 200 lines of the file (security review, PH4 wave 5; server fix in
 *  commit c62756b). */
export interface ExcerptBlock {
  start_line: number;
  lines: string[];
}

export interface SimilarityExcerpt {
  language: string | null;
  /** The same content as `blocks`, flattened to one string with an `...`
   *  separator between non-adjacent blocks. NOT the first 200 lines of the
   *  file — cut around the matched regions instead (±3 lines of context,
   *  merged, capped at 200 lines total). Kept for any caller that only wants
   *  a flat string; screens render from `blocks` (see CodeSimilarityCompare's
   *  `ExcerptPane`), since this string alone carries no line numbers to
   *  highlight against. */
  excerpt: string;
  /** Render from this, not `excerpt` — see `ExcerptBlock`. */
  blocks: ExcerptBlock[];
}

export interface SimilarityCompare {
  signal_id: string;
  low: SimilarityExcerpt;
  high: SimilarityExcerpt;
  /** Empty for a `reference_solution` comparison — the server only pairs
   *  matched regions between two stored fingerprints. */
  matched_regions: MatchedRegion[];
  caption: string;
}

// ---------------------------------------------------------------------------
// Integrity findings — HUMAN evidence, recorded by a named person
// ---------------------------------------------------------------------------

export type FindingOutcome = 'no_concern' | 'follow_up' | 'confirmed';

export interface IntegrityFinding {
  id: string;
  coding_question_id: string;
  signal_id: string | null;
  outcome: FindingOutcome;
  /** `null` once redacted (retention/erasure) — `redacted` says so. */
  rationale: string | null;
  recorded_by_user_id: string;
  created_at: string;
  redacted: boolean;
}

/** The server enforces 20–2000 characters — mirror both bounds so a save
 *  never round-trips into a 422. */
export const FINDING_RATIONALE_MIN = 20;
export const FINDING_RATIONALE_MAX = 2000;

export interface RecordFindingInput {
  attempt_id: string;
  coding_question_id: string;
  outcome: FindingOutcome;
  rationale: string;
  signal_id?: string | null;
  supersedes_id?: string | null;
}

// ---------------------------------------------------------------------------
// Source + evidence bundle
// ---------------------------------------------------------------------------

export interface CodeSource {
  language: string | null;
  /** `null` once redacted — `redacted` says so; never render a placeholder as
   *  if it were the candidate's code. */
  source: string | null;
  redacted: boolean;
}

export interface CodeEvidence {
  attempt_id: string;
  code_redacted: boolean;
  reports: CodeQualityReport[];
  /** The coding round's existing sandbox RESULTS, from `graded_snapshot` —
   *  never coverage. Keyed by `coding_question_id`. */
  test_results: Record<string, CodingResult>;
  integrity: CodeEvidenceIntegrity;
  similarity_signals: AttemptSimilaritySignal[];
  findings: IntegrityFinding[];
}

export interface AnalysisTriggerResult {
  attempts_scanned: number;
  reports_written: number;
  fingerprints_written: number;
  signals_written: number;
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

/** Reports, test results, the integrity summary, similarity signals and
 *  findings for every coding question in this attempt — one call backs the
 *  whole "Code evidence" tab. */
export function getCodeEvidence(examId: string, attemptId: string): Promise<CodeEvidence> {
  return apiGet<CodeEvidence>(
    `/hr/exams/${pathId(examId)}/attempts/${pathId(attemptId)}/code-evidence`,
  );
}

/**
 * The candidate's own submitted source for one coding question — AUDITED ON
 * THE SERVER EVERY TIME THIS IS CALLED. Callers must fetch this only once HR
 * has explicitly asked to see it (a "View source" action), never eagerly
 * alongside the rest of the evidence tab.
 */
export function getCodeSource(
  examId: string,
  attemptId: string,
  codingQuestionId: string,
): Promise<CodeSource> {
  return apiGet<CodeSource>(
    `/hr/exams/${pathId(examId)}/attempts/${pathId(attemptId)}/code/${pathId(codingQuestionId)}`,
  );
}

/** On-demand analysis for one attempt — the sweep already covers recent
 *  submissions; this is for an older one, or a re-run. */
export function triggerCodeAnalysis(
  examId: string,
  attemptId: string,
): Promise<AnalysisTriggerResult> {
  return apiPost<AnalysisTriggerResult>(
    `/hr/exams/${pathId(examId)}/attempts/${pathId(attemptId)}/code-analysis`,
    {},
  );
}

/** Every similarity signal for one exam — counts only are safe to surface on
 *  a list screen; the content lives behind `getSimilarityCompare`. */
/** How much code evidence an application has: counts only -- no source, no
 *  content, no names. Not audited: it reveals nothing about anyone's code. */
export interface CodeEvidenceSummary {
  signal_count: number;
  finding_count: number;
}

export function getCodeEvidenceSummary(enrolmentId: string): Promise<CodeEvidenceSummary> {
  return apiGet<CodeEvidenceSummary>(
    `/hr/enrolments/${pathId(enrolmentId)}/code-evidence-summary`,
  );
}

export function listSimilaritySignals(examId: string): Promise<ExamSimilaritySignal[]> {
  return apiGet<ExamSimilaritySignal[]>(`/hr/exams/${pathId(examId)}/similarity`);
}

/**
 * Excerpts of each matched region, ±3 lines, at most 200 lines per side —
 * AUDITED ON THE SERVER EVERY TIME THIS IS CALLED. Fetch only when HR opens
 * the compare dialog for this specific signal, never ahead of that.
 */
export function getSimilarityCompare(signalId: string): Promise<SimilarityCompare> {
  return apiGet<SimilarityCompare>(`/hr/code-similarity/${pathId(signalId)}`);
}

/** Findings for one attempt, redacted ones excluded by default. `evidence_for_attempt`
 *  already embeds the same (unfiltered) list; this standalone read exists for a
 *  screen that only needs the findings, e.g. after recording a new one. */
export function listIntegrityFindings(
  examId: string,
  attemptId: string,
  includeRedacted = false,
): Promise<IntegrityFinding[]> {
  const q = includeRedacted ? '?include_redacted=true' : '';
  return apiGet<IntegrityFinding[]>(
    `/hr/exams/${pathId(examId)}/attempts/${pathId(attemptId)}/integrity-findings${q}`,
  );
}

// ---------------------------------------------------------------------------
// Writes — HUMAN evidence only
// ---------------------------------------------------------------------------

/** Record a named person's judgement call. `rationale` is mandatory
 *  (20–2000 chars, enforced again by the server) — there is no path to
 *  record a finding without one. */
export function recordCodeIntegrityFinding(body: RecordFindingInput): Promise<{ id: string }> {
  return apiPost<{ id: string }>('/hr/code-integrity-findings', body);
}
